rust/xaynet-client/src/mobile_client/participant/mod.rs

```diff
@@ -64,14 +64,15 @@ pub struct Participant<Task> {
 }
 
 impl<Task> Participant<Task> {
-    /// Sign the given message with the participant secret key, and
-    /// encrypt the signed message with the given public key.
+    /// Serialize, sign and encrypt the given message.
+    ///
+    /// The message is signed with the participant secret signing
+    /// key. `pk` is the coordinator public key, used to encrypt the
+    /// final message
     pub fn seal_message(&self, pk: &CoordinatorPublicKey, message: &Message) -> Vec<u8> {
-        let message_seal = MessageSeal {
-            recipient_pk: pk,
-            sender_sk: &self.state.keys.secret,
-        };
-        message_seal.seal(message)
+        let mut buf = vec![0; message.buffer_length()];
```

Comment:
Does it makes sense to add the `seal_message` method to `Message`? For me it feels a bit low level to work with bytes of the message.

Reply:
I considered it but I had to make a choice between two APIs, but felt that the trade-offs should be made by the caller instead:

```
impl Message
  // This allocates two vec: one for serializing the message, one for the encrypted data. It's user friendly, and can't panic
  fn seal_1(&self, sk: SecretSigningKey, pk: PublicEncryptKey) -> Vec<u8> {
    let mut buf = vec![0; self.buffer_length()];
    self.to_bytes(&mut buf, sk);
    pk.encrypt(&buf[..])
  }

  // This allocates just one vec for the encrypted data. It's more consistent with the existing to_bytes method but it can panic.
  fn seal_2(&self, buf: &mut [u8], sk: SecretSigningKey, pk: PublicEncryptKey) -> Vec<u8> {
    self.to_bytes(&mut buf, sk);
    pk.encrypt(&buf[..])
  }
```

Apart from that, the caller may decide to do encryption differently (using the new [streaming API](https://docs.rs/sodiumoxide/0.2.6/sodiumoxide/crypto/secretstream/index.html) for instance). So in the end, implementing that method here means we'd have to take some decisions I'd prefer the caller to take.
-